iree/compiler/Codegen/LLVMGPU/LLVMGPULowerExecutableTarget.cpp

```diff
@@ -74,45 +88,74 @@ void LLVMGPULowerExecutableTargetPass::runOnOperation() {
+  Optional<mlir::ArrayAttr> workgroupSizeAttr;
   for (auto &it : entryPoints) {
     auto entryPointOp = it.second;
-    if (IREE::Codegen::TranslationInfoAttr translationInfo =
+    if (IREE::Codegen::TranslationInfoAttr currTranslationInfo =
             getTranslationInfo(entryPointOp)) {
-      IREE::Codegen::DispatchLoweringPassPipeline currPipeline =
-          translationInfo.getDispatchLoweringPassPipeline();
-      if (passPipeline) {
-        if (currPipeline != passPipeline.getValue()) {
-          moduleOp.emitError(
-              "unhandled compilation of entry point function with different "
-              "pass pipelines within a module");
-          return signalPassFailure();
+      if (translationInfo) {
+        if (currTranslationInfo != translationInfo.getValue()) {
+          moduleOp.emitOpError(
+              "unhandled compilation of entry point functions with different "
+              "translation info");
         }
-        continue;
+      } else {
+        translationInfo = currTranslationInfo;
       }
-      passPipeline = currPipeline;
+      // There should only be one entry op and workgroup size for now.
+      workgroupSizeAttr = entryPointOp.workgroup_size();
     }
   }
 
-  executableLoweringPipeline.addPass(createSetNumWorkgroupsPass());
-  executableLoweringPipeline.addPass(createCanonicalizerPass());
-  if (!testLoweringConfiguration && passPipeline.hasValue()) {
-    OpPassManager &nestedModulePM = executableLoweringPipeline.nest<ModuleOp>();
-    switch (*passPipeline) {
+  // Verify the configuration.
+  if (translationInfo.hasValue() && workgroupSizeAttr.hasValue()) {
+    std::array<int64_t, 3> workgroupSizes;
+    for (auto it : llvm::enumerate(workgroupSizeAttr.getValue())) {
+      workgroupSizes[it.index()] = it.value().cast<IntegerAttr>().getInt();
+    }
+
+    LogicalResult verificationStatus = success();
+    switch (translationInfo.getValue().getDispatchLoweringPassPipeline()) {
       case IREE::Codegen::DispatchLoweringPassPipeline::LLVMGPUDistribute:
-        addGPUSimpleDistributePassPipeline(nestedModulePM);
         break;
       case IREE::Codegen::DispatchLoweringPassPipeline::LLVMGPUVectorize:
-        addGPUVectorizationPassPipeline(nestedModulePM);
         break;
       case IREE::Codegen::DispatchLoweringPassPipeline::LLVMGPUMatmulSimt:
-        addGPUMatmulSimtPassPipeline(nestedModulePM);
+        verificationStatus = verifyLoweringConfiguration(
+            moduleOp, translationInfo.getValue(), workgroupSizes,
+            verifyGPUMatmulSimtPassPipeline);
         break;
       case IREE::Codegen::DispatchLoweringPassPipeline::LLVMGPUMatmulTensorCore:
-        addGPUMatmulTensorCorePassPipeline(nestedModulePM);
         break;
-      default:
-        llvm_unreachable("Unsupported pipeline on GPU target.");
+      default:;
+    }
+    if (failed(verificationStatus)) {
+      return signalPassFailure();
+    }
```

Comment:
Can we move all this logic in a separate function? We could do verification for each entry point? (even though right now there should be only one entry point

Reply:
well if you expect only one entry point, you should fail if there are more than one entry points. Right now its #yolo -ing and assuming only one entry point.

Reply:
Should I explicitly verify there is only one entry point?

Reply:
I think we can just verify each entry point, I don't see why this is more complicated?